src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2012,8 +2012,13 @@ internal void MarkStaticConstructorVisibleToReflection (TypeDefinition type, in
 
 			var typeOrigin = new MessageOrigin (type);
 
-			foreach (Action<TypeDefinition> handleMarkType in MarkContext.MarkTypeActions)
-				handleMarkType (type);
+			if (MarkContext.MarkTypeActions.Count > 0) {
+				foreach (Action<TypeDefinition> handleMarkType in MarkContext.MarkTypeActions)
+					handleMarkType (type);
+
+				// Rebuild type info for the type in case a mark action added new methods.
+				Annotations.TypeMapInfo.MapType (type);
```

Comment:
Does this mean that if the custom step registers a type action then we will recreate the mapinfo for all types we mark... well I guess we don't have a way to tell if the type action actually did anything... it's unfortunate.

Reply:
If we want to avoid significant perf degradation (and stability issues), I think we need to be able to guarantee the type map is read/init only. Type mapping is one of the more expensive parts of the linker. Maybe we could consider enforcing that no changes to IL can occur after MarkStep has started? Can the custom step with the issue run before MarkStep? It doesn't look like it depends on any marking info.

Reply:
> that no changes to IL can occur after MarkStep has started

The trimmer itself violates this - the constant propagation and branch elimination happens during method processing in MarkStep - this was one of the big perf optimizations we've done in previous releases, to avoid running the constant prop on all methods in the input and instead only run it on "marked" methods. (This is a lot more complicated and the description is a bit fuzzy, but the core is true).

Re custom steps: I think this is similar - we want to optimize and only run the custom step on marked types, so that we avoid doing work for types which are going to be trimmed away. I don't know if there would be side effects if we did that.

We could do something similar to what the constant-prop does for methods - it also doesn't run on JUST marked methods. Instead, it starts from marked methods but then follows the callgraph to any method it needs to go to. The type map could do something similar, but it would be tricky to get it right - for example if we blindly collect all implementing methods for some core interface (like IEnumerable), we would end up processing the entire input. But then we could wire the custom step to run as part of the type map instead, so it would run "before" the type map sees the IL of a given type. It would not be the same as now - the custom step would run on more types, but hopefully still better then it running on everything. (and we would still have to check that it's OK to run the custom step on more types than necessary).

Overall I'm fine with the current behavior - the cost is only incurred for trimmer runs which use custom steps (so basically mono only) - and for a good reason (mono needs the custom steps currently).

I hope to get to do more work on mono interop in .NET 10 and hopefully truly start moving away from custom steps. But I expect that we will need to keep them around for a couple of years still (even if we figure out the correct solution, it's unlikely we would be able to adapt it for both iOS and Android in one release).